proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

```diff
@@ -0,0 +1,243 @@
+foo.methodOnCompleteFoo() // error no member 'methodOnCompleteFoo' on type 'IncompleteFoo.Foo'
+```
+
+```swift
+import IncompleteFoo
+import CompleteFoo
+let completeFoo = Foo() // This time it works because the cache was not populated until after CompleteFoo was imported
+foo.methodOnCompleteFoo()  // Works as well
+```
+
+This happens because `ClangImporter` is lazy and caches imports, once the placeholder is synthesized (when only the incomplete definition is
+available), the complete definition will never be imported. Without an understanding of `ClangImporter`, it is impossible to understand why
+the statements between the two imports have an effect on those after. We have made attempts at solving these issues, but no satisfactory solution
+was found. The details of these attempts are listed in the alternatives considered section. 
+
+The [existing ClangImporter diagnostics](https://forums.swift.org/t/pitch-improved-clangimporter-diagnostics/52687) will still properly warn users about forward declaration limitations, with or with out the feature enabled.
+
+## Source compatibility
+
+This change does have the potential to break source compatibility, since
+`ClangImporter` will now import declarations that were previously dropped.
+That said, the impact should rare. The current implementation already passes
+the source compatibility suite and the source kit stress test.
+
+As mentioned, to address source compatibility issues, the feature is gated behind
+a flag for Swift 5 and before, and enabled by default in Swift 6.
+
+## Effect on ABI stability
+
+This change has no affect on ABI.
+
+## Effect on API resilience
+
+This change does not influence the ABI characteristics
+of public APIs.
+
+## Alternatives considered
+
+The unavailable attribute attached to synthesized declarations might be more restrictive than necessary.
+We considered the introduction of a new attribute to denote an incomplete type. This attribute would be used to prevent
+operations that are not suitable for an incomplete type (ie. declaring a Swift type that conforms to an incomplete protocol).
+
+Allowing more use of the type however increases the complexity of the problem, and the likelihood of source compatibility issues.
+For example, if the synthesized type is not file private, issues could arise with different files having different representations
+of the type depending on which modules they import.
+
+We also attempted to make this implementation usable from the REPL. As mentioned, we had issues with `ClangImporter` caching
+imports. We must invalidate the cache, at the very least to allow the new definition of type 'Foo' to be imported. This creates issues
+where two incompatible representations of the type (`IncompleteFoo.Foo` and `CompleteFoo.Foo`) are in context at the same time.
+To solve these issues we attempted the following:
+
+1. Invalidate all direct or indirect references to `IncompleteFoo.Foo` and replace them with `CompleteFoo.Foo`:
```

Comment:
Let me know if this is too implementation detail heavy. The TLDR is the REPL problem is technically difficult to crack, more difficult to crack than the non-REPL problem -- at least with my knowledge.